### PR TITLE
Harmonize `AsyncFn` implementations, make async closures conditionally impl `Fn*` traits

### PR DESCRIPTION
**Pull request by compiler-errors**
This PR implements several changes to the built-in and libcore-provided implementations of `Fn*` and `AsyncFn*` to address two problems:
1. async closures do not implement the `Fn*` family traits, leading to breakage: https://crater-reports.s3.amazonaws.com/pr-120361/index.html
2. *references* to async closures do not implement `AsyncFn*`, as a consequence of the existing blanket impls of the shape `AsyncFn for F where F: Fn, F::Output: Future`.

In order to fix (1.), we implement `Fn` traits appropriately for async closures. It turns out that async closures can:
* always implement `FnOnce`, meaning that they're drop-in compatible with `FnOnce`-bound combinators like `Option::map`.
* conditionally implement `Fn`/`FnMut` if they have no captures, which means that existing usages of async closures should *probably* work without breakage (crater checking this: https://github.com/rust-lang/rust/pull/120712#issuecomment-1930587805).

In order to fix (2.), we make all of the built-in callables implement `AsyncFn*` via built-in impls, and instead adjust the blanket impls for `AsyncFn*` provided by libcore to match the blanket impls for `Fn*`.

### CODE REVIEW

**Comment by compiler-errors** [comment]
@craterbot check crates=https://crater-reports.s3.amazonaws.com/pr-120361/retry-regressed-list.txt p=1

**Comment by compiler-errors** [comment]
Hey CI, wtf is wrong with you?

**Comment by compiler-errors** [comment]
r? oli-obk

**Comment by compiler-errors** [comment]
idk what's up with CI but this PR is at least ready for review, let me know if there are any questions

File: compiler/rustc_hir_typeck/src/callee.rs [hunk 1]
```diff
@@ -183,16 +183,23 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     kind: TypeVariableOriginKind::TypeInference,
                     span: callee_expr.span,
                 });
+                // We may actually receive a coroutine back whose kind is different
+                // from the closure that this dispatched from. This is because when
+                // we have no captures, we automatically implement `FnOnce`. This
+                // impl forces the closure kind to `FnOnce` i.e. `i8`.
```
Comment (oli-obk) [inline compiler/rustc_hir_typeck/src/callee.rs:189]: uh... explain that `i8` comment please ^^

**Review by oli-obk — CHANGES_REQUESTED** [review]

File: compiler/rustc_trait_selection/src/traits/project.rs [hunk 2]
```diff
@@ -2456,76 +2457,48 @@ fn confirm_async_closure_candidate<'cx, 'tcx>(
...
-                        ty::GenericArg::from(kind_ty),
-                        Ty::from_closure_kind(tcx, goal_kind).into(),
-                        env_region.into(),
-                        sig.tupled_inputs_ty.into(),
-                        args.tupled_upvars_ty().into(),
-                        args.coroutine_captures_by_ref_ty().into(),
-                    ],
-                );
-                let coroutine_ty = sig.to_coroutine(
+    let item_name = tcx.item_name(obligation.predicate.def_id);
```
[the remark this answers is not in the document]
Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/project.rs:2526]: There are no lang items for these associated types

File: compiler/rustc_hir_typeck/src/callee.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_hir_typeck/src/callee.rs:189]: This is the correspondence between Ty and ClosureKind: https://doc.rust-lang.org/stable/nightly-rustc/src/rustc_middle/ty/closure.rs.html#131

but I got it backwards whoops

Comment (compiler-errors) [reply compiler/rustc_hir_typeck/src/callee.rs]: I ended up removing this hack in favor of another one that's a bit easier to understand: 3bb384aad6e7f61a0b4b8c604206e78ffa418df4

**Review by oli-obk — APPROVED** [review]
r=me (needs blessing)

**Comment by compiler-errors** [comment]
waiting on https://github.com/rust-lang/rust/pull/120812 i think

**Comment by compiler-errors** [comment]
Actually, let's land with this type name sorting hack (the final commit) for now. 

I will revert it in #120812, which @cjgillot says may take until next week to land, but I'd rather not block async closure experimentation on a single diagnostic bug.

**Comment by compiler-errors** [comment]
either that, or I can ignore the stderr on `tests/ui/async-await/async-fn/dyn-pos.rs`, or normalize it away. I am open to suggestions

**Comment by oli-obk** [comment]
Diagnostic sorting seems fine to me

@bors r+

**Comment by compiler-errors** [comment]
That CI failure looks totally unrelated

@bors retry

**Comment by compiler-errors** [comment]
Linker error on clif looks unrelated?

@bors retry

**Pull request merged**